test/nested_derivs_unit.C

```diff
@@ -0,0 +1,37 @@
+#include "metaphysicl/dualnumberarray.h"
+
+using namespace MetaPhysicL;
+typedef double Real;
+
+typedef DualNumber<Real, NumberArray<4, Real>> ADReal;
+typedef DualNumber<ADReal, NumberArray<2, ADReal>> CompoundADType;
+
+int
+main(int argc, const char * argv[])
+{
+  ADReal u0 = 2;
+  u0.derivatives()[0] = 1;
+  ADReal u1 = 3;
+  u1.derivatives()[1] = 1;
+  ADReal v0 = 4;
+  v0.derivatives()[2] = 1;
+  ADReal v1 = 5;
+  v1.derivatives()[3] = 1;
+
+  ADReal u = 0.5 * (u0 + u1);
+  u = u + 0;
+  ADReal v = 0.5 * (v0 + v1);
+
+  const ADReal zero(0, 0);
+  const ADReal one(1, 0);
+
+  CompoundADType u_c(u, zero);
+  auto & u_cd = u_c.derivatives();
+  u_cd[0] = one;
+
+  CompoundADType v_c(v, zero);
+  auto & v_cd = v_c.derivatives();
+  v_cd[1] = one;
+
+  u_c = u_c + zero;
```

Comment:
@roystgnr what I want to happen here is for `zero` to essentially get auto-promoted to a `CompoundADType` with `zero` as the `value()` and `derivatives()` corresponding to `NumberArray<2, ADReal>{}`. However, what happens without https://github.com/libMesh/MetaPhysicL/pull/36/commits/fde3fb7d47a5cb261d744e7f2981f11588147ebc is that we attempt to add the `derivatives()` of an `ADReal` with the `derivatives()` of the `CompoundADType`. This kind of thing will happen over and over for `DualNumber<T,D,asd>` and `DualNumber<T2,D2,asd>` operands, even though here `T = DualNumber<T2,D2,asd>`. With https://github.com/libMesh/MetaPhysicL/pull/36/commits/fde3fb7d47a5cb261d744e7f2981f11588147ebc I make some progress but eventually run into the most epic block of template error message I've ever seen. A snippet:
```
../../test/navier_unit.h:119:20:   required from here
../../src/core/include/metaphysicl/compare_types.h:217:37: error: ambiguous template instantiation
for ‘struct MetaPhysicL::CompareTypes<MetaPhysicL::DualNumber<MetaPhysicL::DualNumber<double,
MetaPhysicL::SparseNumberStruct<MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<0, double>,
MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<1, double>,
MetaPhysicL::NullContainer<MetaPhysicL::UnsignedIntType<1, double> >, MetaPhysicL::ValueLessThan>,
MetaPhysicL::ValueLessThan> >, false>,
MetaPhysicL::SparseNumberStruct<MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<0,
MetaPhysicL::DualNumber<double, MetaPhysicL::SparseNumberStruct<MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<0,
double>, MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<1, double>,
MetaPhysicL::NullContainer<MetaPhysicL::UnsignedIntType<0, double> >,
MetaPhysicL::ValueLessThan>, MetaPhysicL::ValueLessThan> >, false> >,
MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<1, MetaPhysicL::DualNumber<double,
MetaPhysicL::SparseNumberStruct<MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<0, double>,
MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<1, double>, MetaPhysicL::NullContainer<MetaPhysicL::UnsignedIntType<1,
double> >, MetaPhysicL::ValueLessThan>, MetaPhysicL::ValueLessThan> >, false> >,
MetaPhysicL::NullContainer<MetaPhysicL::UnsignedIntType<1, MetaPhysicL::DualNumber<double,
MetaPhysicL::SparseNumberStruct<MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<1, double>,
MetaPhysicL::NullContainer<MetaPhysicL::UnsignedIntType<1, double> >, MetaPhysicL::ValueLessThan> > > > >,
MetaPhysicL::ValueLessThan>, MetaPhysicL::ValueLessThan> >, false>, MetaPhysicL::DualNumber<double,
MetaPhysicL::SparseNumberStruct<MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<0, double>,
MetaPhysicL::Container<MetaPhysicL::UnsignedIntType<1, double>, MetaPhysicL::NullContainer<MetaPhysicL::UnsignedIntType<1,
double> >, MetaPhysicL::ValueLessThan>, MetaPhysicL::ValueLessThan> >, false>, false, void>’
```
Is there a better way to get what I'm hoping for here? You have so many intricate unit tests in the repo; I have some belief that you've done things like this somewhere in there. With the `GeomReal` work in libMesh/MOOSE I would really like to have the same kind of type-promotion of `Real -> ADReal` for `ADReal -> CompoundADType` or else some of the programmed Newton loops in mortar, which leverage automatic differentiation to determine projections, will need more manual massaging.

Reply:
So we'd want the output to be a `TypeVector<DualNumber<ADReal, NumberArray<2, ADReal>>` in that case, yes?

I'm kind of surprised that that's not working, if the corresponding `TypeVector<Real> * DualNumber<ADReal, NumberArray<2, ADReal>>` is working.  The difficulty with these mixed upconversions was always in making sure there was a `supertype` metafunction to tell the compiler which nested template class should be the "outer" nesting (so it knows not to try to produce a `DualNumber<TypeVector<ADReal>, TypeVector<NumberArray<2, ADReal>>>` or a `DualNumber<TypeVector<ADReal>, NumberArray<2, TypeVector<ADReal>>>` instead), but if we've already got `TypeVector` wrapping outside `DualNumber` then I'd have expected the whole thing to work out.

Do we have any unit test coverage anywhere for the `Point` (of `Real`) ` * ADReal` case?

Reply:
The thing is that this doesn't currently work according to my desire:

DualNumber<DualNumber<T, D, asd>, DD, asd> * DualNumber<T, D, asd>

so I have little hope that this will work:

DualNumber<DualNumber<T, D, asd>, DD, asd> * TypeVector<DualNumber<T, D, asd>>

Reply:
The result of

DualNumber<DualNumber<T, D, asd>, DD, asd> * DualNumber<T, D, asd>

is

DualNumber<typename CompareTypes<DualNumber<T, D, asd>, T>::supertype,
                      typename CompareTypes<DD, D>::supertype,
                      asd>

let's say for the sake of argument that I only had one Newton deriv, I wanted to take, so actually DD is another ADReal. Then I am taking the supertype comparing types of an ADReal and likely a big long container of just Reals. Consequently I wind up something like (the container type is wrong, but this is just illustrative):

DualNumber<DualNumber<T, D, asd>,
                       NumberArray<50, DualNumber<T, D, asd>>,
                       asd>

when I wanted just

DualNumber<DualNumber<T, D, asd>
                       DualNumber<T, D, asd>
                       asd>

Reply:
This is probably not the best venue for this discussion as pull request conversations do not automatically update when someone comments so I have to refresh my browser page if I want to see if you've commented...

Reply:
> so I have to refresh my browser page if I want to see if you've commented

Not the best interface, no.  I do like this stuff being archived more publicly than Slack or conversations, but maybe it would be better to hash things out with less latency and then just archive the conclusions.

> `DualNumber<DualNumber<T, D, asd>, DD, asd> * DualNumber<T, D, asd>`

This should have been working, at least in the dumb "T is a scalar, D is a vector of scalars, DD is a vector of first-level DualNumbers" case.  That was the original use case for MetaPhysicL: feeding manufactured solutions through second-order PDEs in order to get forcing functions.  I'm going to kick myself so hard if somehow we've disabled the test coverage and it's regressed.  I believe we had test coverage for it in navier_unit.C?

Reply:
Okay, I've been trying, maybe my brain's just trying to take an early weekend, but I honestly don't understand this case:

> `DualNumber<DualNumber<T, D, asd>, DD, asd>`

Let's use `DN` (and `R` for `Real`, `NA` for `NumberArray<50...`) for shorthand, ignore `asd`; then with `D=NA<R>` and `DD=DN<T,D>` this expands out to be:

`DN<DN<R,NA<R>>,DN<R,NA<R>>>`

... and then if we were to take `derivatives().value()` we'd have a single scalar, but if we took `value().derivatives()` we'd have an array.  That can't be a consistent representation of a value with derivatives.  I've never been thrilled with the way simply nesting `DualNumber` represents higher derivatives, since they store symmetric indices redundantly and since someone could in theory make a non-symmetric modification to an index at run-time and come up with something that makes no sense ... but this doesn't even make sense at compile-time with first derivatives, does it?